Mathlib/AlgebraicGeometry/Over.lean

```diff
@@ -0,0 +1,94 @@
+/-
+Copyright (c) 2024 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+import Mathlib.AlgebraicGeometry.Scheme
+
+/-!
+# Typeclasses for `S`-schemes and `S`-morphisms
+
+## Main definition
+- `AlgebraicGeometry.Scheme.Over`: `X.Over S` equips `X` with a `S`-scheme structure.
+  `X ⮕ S : X ⟶ S` is the structure morphism.
+- `AlgebraicGeometry.Scheme.Hom.IsOver`: `f.IsOver S` asserts that `f` is a `S`-morphism.
+
+-/
+
+
+namespace AlgebraicGeometry.Scheme
+
+universe u
+
+open CategoryTheory
+
+variable {X Y : Scheme.{u}} (f : X.Hom Y) (S S' : Scheme.{u})
+
+/--
+`X.Over S` is the typeclass containing the data of a structure morphism `X ⮕ S : X ⟶ S`.
+-/
+protected class Over (X S : Scheme.{u}) : Type u where
+  ofHom ::
+  /-- The structure morphism. Use `X ⮕ S` instead. -/
+  hom : X ⟶ S
```

Comment:
I think we should define this for an arbitrary category `C` instead and probably name it `IsOver`. If we want the nice `X.Over S` notation, we can have an `abbrev` for `Scheme.Over`.

Reply:
Well, I think any geometric theory building upon category theory will encounter the same issue, e.g. adic spaces.

Reply:
Hmmmm. I'll move this to category with a strong warning then.